internal/core/component_factory.rs

```diff
@@ -0,0 +1,287 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+#![warn(missing_docs)]
+
+//! This module contains the `ComponentFactory` and a `Component`. Both are
+//! exposed to the C API.
+
+use core::ffi::c_void;
+use core::fmt::Debug;
+
+use alloc::{boxed::Box, rc::Rc};
+
+use crate::{
+    api::ComponentHandle,
+    component::{ComponentRc, ComponentVTable},
+};
+
+/// A `Component` built by a `ComponentFactory`
+#[repr(C)]
+pub struct Component(ComponentRc);
+
+impl<T: ComponentHandle> From<T> for Component
+where
+    T::Inner: vtable::HasStaticVTable<ComponentVTable> + 'static,
+{
+    fn from(value: T) -> Self {
+        Self(vtable::VRc::into_dyn(value.as_weak().inner.upgrade().unwrap()))
+    }
+}
+
+impl Component {
+    pub(crate) fn component_rc(self: Self) -> ComponentRc {
+        self.0
+    }
+
+    pub(crate) fn clone_internal(&self) -> Component {
+        Component(self.0.clone())
+    }
+}
+
+pub(crate) mod internal {
+    use super::{c_void, Component, Rc};
+
+    /// A internal helper needed to implement `ComponentFactory` in different languages.
+    #[derive(Clone, PartialEq)]
+    #[repr(C)]
+    pub struct ComponentFactoryData {
+        pub build: unsafe extern "C" fn(*mut c_void) -> *mut Component,
+        pub data: *mut c_void,
+        pub drop: unsafe extern "C" fn(*mut c_void),
+    }
+
+    #[allow(unsafe_code)]
+    impl ComponentFactoryData {
+        pub fn build(&self) -> Option<Component> {
+            let result = unsafe { (self.build)(self.data) };
+            if result.is_null() {
+                None
+            } else {
+                let r = unsafe { (*result).clone_internal() };
+                Some(r)
+            }
+        }
+    }
+
+    impl Default for ComponentFactoryData {
+        fn default() -> Self {
+            let data = Box::new(RustData { function: Box::new(|| None) });
+            Self {
+                build: __slint_component_factory_builder_call,
+                data: Box::into_raw(data) as *mut c_void,
+                drop: __slint_component_factory_builder_drop,
+            }
+        }
+    }
+
+    #[allow(unsafe_code)]
+    impl Drop for ComponentFactoryData {
+        fn drop(&mut self) {
+            unsafe { (self.drop)(self.data) };
+        }
+    }
+
+    pub(crate) struct RustData {
+        pub(crate) function: Box<dyn FnMut() -> Option<Component>>,
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) unsafe extern "C" fn __slint_component_factory_builder_drop(data: *mut c_void) {
```

Comment:
Why the funny name? (starting with __slint)

Not wise to start with underscore otherwise we wouldn't get the unused warning if it's not used.